File: src/python/components/event.py
from datetime import datetime

from src.python.components.player import Player

from src.python.helpers.map import M

class Event:
    def __init__(self, persona):
        self.persona = persona

    def dispatch(self, event_key, content=None):
        now_in_local_timezone = datetime.now().astimezone()

        output = now_in_local_timezone.replace(microsecond=0).isoformat()

        if M.dig(self.persona, 'meta', 'symbol'):
            output += ' ' + self.persona['meta']['symbol']

        output += ' >'

        if not M.dig(self.persona, 'events', event_key):
            output += ' [' + event_key + ']'
            if content:
                output += ' ' + content
            print(output)
            return

        if M.dig(self.persona, 'events', event_key, 'symbol'):
            output += ' ' + self.persona['events'][event_key]['symbol']
        else:
            output += ' [' + event_key + ']'
        
        if M.dig(self.persona, 'events', event_key, 'message'):
            output += ' ' + self.persona['events'][event_key]['message']

        if content:
            output += ' ' + content

        print(output)

        if M.dig(self.persona, 'events', event_key, 'audio'):
            Player.play(
                self.persona['events'][event_key]['audio'],
                M.dig(self.persona, 'events', event_key, 'volume') or 1)

File: src/ruby/components/event.rb
# frozen_string_literal: true

require 'time'

require './src/ruby/components/player'

class Event
  def initialize(persona)
    @persona = persona
  end

  def dispatch(event_key, content = nil, silent = false)
    event_key = event_key.to_sym

    output = Time.now.iso8601.to_s

    output += " #{@persona[:meta][:symbol]}" if @persona.dig(:meta, :symbol)

    output += ' >'

    unless @persona.dig(:events, event_key)
      output = "#{output} [#{event_key}]"
      output += " #{content}" if content
      puts output
      return
    end

    event = @persona[:events][event_key]

    output += if event[:symbol] == 'nano-bot-cartridge' && @persona.dig(:cartridge, :meta, :symbol)
                " #{@persona[:cartridge][:meta][:symbol]}"
              elsif event[:symbol]
                " #{event[:symbol]}"
              else
                " [#{event_key}]"
              end

    output += " #{event[:message]}" if event[:message]

    output += " #{content}" if content

    puts output

    Player.play!(event[:audio], event[:volume] || 1) if event[:audio] && !silent
  end
end
